torch/csrc/jit/init.cpp

```diff
@@ -122,7 +122,12 @@ void initJITBindings(PyObject* module) {
           [](std::shared_ptr<Graph>& g) { return PropagateQuantInfo(g); })
       .def(
           "_jit_pass_insert_observers",
-          [](std::shared_ptr<Graph>& g) { return InsertObserverNodes(g); })
+          [](std::shared_ptr<Graph>& g, py::function pp) {
+            Node* new_node =
```

Comment:
ok, so you use "name" as identity, that works too. Another option I was playing with is to pass into a "class" instance and create new instance for each observer not we insert. This way we don't need to rely on any names at all (other for debugging): https://github.com/pytorch/pytorch/compare/master...dzhulgakov:observer-hack-illustration?expand=1 @nishantpdce

Reply:
Right. I think both options are almost equivalent. The advantage of this approach would be:
- Potential better debugging experience, as you mentioned,
- It's more transparent to the user,
- Less coupling between passes: I can remove observers and save the model, then I can load it and immediately quantize. Alternatively, I can quantize even without running the observers if I prepare the data myself in the correct format,
- It's already done here :)

PS: Also, thanks for commenting on exactly this line of code - I noticed some pieces that I forgot to cleanup!

Reply:
Regarding loading - do we guarantee preservation of unique_name through serialization?

The question is whether it's safe to rely on uniqueName being stable between passes? The benefit of explicit objects is that it'd preserve identity of an operator explicitly, even with nested blocks/modules. E.g. imagine that you're going to start quantizing a set of several nested modules that can call each other. You'd need to start passing method and module names as part of the fully qualified name too. Or we'd be quantizing on per function basis only?

In general I agree they are equivalent at this point.

Reply:
There are no guarantees regarding the names, but it shouldn't be an issue since we embed the actual name as a constant parameter into every call to the observer function. Thus, when we use the dictionary, the key would be that second argument, not the unique name of the value. In practice, they would most probably match, but there are no guarantees on that.

As for the nested submodules, I expect we might need to refine the way we construct the key in the dictionary (e.g. use "scope:unique_name" instead of just "unique_name"), but overall the approach should still work.

Note btw that this approach can be used for inserting any type of observers, not just for quantization observers. E.g. if the observer function records timing since its last invocation, you would get a time profile for free.

Reply:
Are names preserved across passes? For example, if we have passes between observer and quantizer, can we assume that the same names exist?

Reply:
It would be good to run this through with simple LeNet (script and trace) just to make sure all nodes are visited and get the observer

Reply:
Names are note guaranteed to be preserved across passes, however this approach would still work correctly because we embed the names into the observer calls.

The correspondence between the old observed stats and a value would be restored from the arguments of the observer call and will not depend on the name of the value. E.g. in
`x1 = observe(x, "some_name")
`
We know that `value_stats["some_name"]` ("some_name" is the 2nd argument) contains observed stats for `x` (`x` is the 1st argument).

Note that we can use any other identifier instead of the name here, and it would work just as well, using the name just gives us some inspectability for free.